src/core/app.lua

```diff
@@ -119,66 +122,138 @@ end
+         add_action('link_output', fa, fl, linkspec)
+      end
+      if fresh_link or fresh_apps[ta] then
+         add_action('link_input', ta, tl, linkspec)
+      end
+   end
+
    return actions
 end
 
 -- Update the active app network by applying the necessary actions.
-function apply_config_actions (actions, conf)
-   -- The purpose of this function is to populate these tables:
-   local new_app_table,  new_app_array  = {}, {}
-   local new_link_table, new_link_array = {}, {}
-   -- Temporary name->index table for use in link renumbering
-   local app_name_to_index = {}
+function apply_config_actions (actions)
    -- Table of functions that execute config actions
    local ops = {}
-   function ops.stop (name)
-      if app_table[name].stop then
-         app_table[name]:stop()
-      end
-      if app_table[name].shm then
-         shm.delete_frame(app_table[name].shm)
+   -- As an efficiency hack, some apps rely on the fact that we add
+   -- links both by name and by index to the "input" and "output"
+   -- objects.  Probably they should be changed to just collect their
+   -- inputs and outputs in their :link() functions.  Until then, call
+   -- this function when removing links from app input/output objects.
+   local function remove_link_from_array(array, link)
+      for i=1,#array do
+         if array[i] == link then
+            table.remove(array, i)
+            return
+         end
       end
    end
-   function ops.keep (name)
-      new_app_table[name] = app_table[name]
-      table.insert(new_app_array, app_table[name])
-      app_name_to_index[name] = #new_app_array
+   function ops.unlink_output (appname, linkname)
+      local app = app_table[appname]
+      local link = app.output[linkname]
+      app.output[linkname] = nil
+      remove_link_from_array(app.output, link)
+      if app.link then app:link() end
+   end
+   function ops.unlink_input (appname, linkname)
+      local app = app_table[appname]
+      local link = app.input[linkname]
+      app.input[linkname] = nil
+      remove_link_from_array(app.input, link)
+      if app.link then app:link() end
+   end
+   function ops.free_link (linkspec)
+      link.free(link_table[linkspec], linkspec)
+      link_table[linkspec] = nil
+      configuration.links[linkspec] = nil
+   end
+   function ops.new_link (linkspec)
+      link_table[linkspec] = link.new(linkspec)
+      configuration.links[linkspec] = true
+   end
+   function ops.link_output (appname, linkname, linkspec)
+      local app = app_table[appname]
+      local link = assert(link_table[linkspec])
+      app.output[linkname] = link
+      table.insert(app.output, link)
+      if app.link then app:link() end
+   end
+   function ops.link_input (appname, linkname, linkspec)
+      local app = app_table[appname]
+      local link = assert(link_table[linkspec])
+      app.input[linkname] = link
+      table.insert(app.input, link)
+      if app.link then app:link() end
```

Comment:
No, checking for link changes in a pull function is a big caveat imho, it would make a the respective pull callbacks ugly.

Its kind of mind bending to me that `Follower` and `Leader` are supposed to be apps, instead of programs. To me it seems simpler to stream configuration changes in between calls to `enginine.main`, and only apply them once they are complete.